### PR TITLE
added the solar time

### PR DESCRIPTION
**Pull request by SallyElHajjar**
I have added the solar time and tested it in my local.  I have changed the expected values to remove the errors. Could you confirm ? @MadhaviJain90

### CODE REVIEW

File: usl_pipeline/cloud_functions/main.py [hunk 1]
```diff
@@ -1027,6 +1028,74 @@ def _direction_to_cosine(degrees):
     return dataset
 
 
+def _compute_solar_time_components(dataset: xarray.Dataset) -> xarray.Dataset:
+    # Solar Time Computation for every Longitude of the City
+    # Solar Time Sine and Cosine Derivation from there
+    if all(var in dataset.keys() for var in ["XLONG_M", "XLAT_M", "Times"]):
```
Comment (Katsutoshii) [inline usl_pipeline/cloud_functions/main.py:1034]: How should we handle the case when these features are missing? Should we report an error to the user?

**Review by Katsutoshii — APPROVED** [review]

File: usl_pipeline/cloud_functions/main.py [hunk 2]
```diff
@@ -1027,6 +1028,81 @@ def _direction_to_cosine(degrees):
...
+
+        # Extract hours and minutes from the datetime objects
+        utc_hours_minutes = (
+            times.astype("datetime64[h]").astype(int) % 24
+            + times.astype("datetime64[m]").astype(int) % 60 / 60
+        )
+
+        # Define the function to calculate solar time
+        def calculate_solar_time(utc_time, longitude):
+            return (utc_time + longitude / 15) % 24
```
Comment (ahmustafa) [inline usl_pipeline/cloud_functions/main.py:1065]: what if we have negative values of longitude (west of the prime meridian)?
we may adjust as 
return ((utc_time + longitude / 15) % 24 + 24) % 24

Comment (MadhaviJain90) [reply usl_pipeline/cloud_functions/main.py:1065]: For negative longitudes like NYC the code will take care of it. Eg, if WRF variable XLONG_M = -75, then local time would be UTC-(5) hrs. What I am concerned about is the sin and cos solar time values that I was getting in my code and the values that the unit tests in this one is giving

**Review by Katsutoshii — APPROVED** [review]

**Comment by MadhaviJain90** [comment]
I don't see any differences from the Solar Time PR I had coded -> https://github.com/UrbanSystemsLab/climateiq-cnn/pull/170, except a minor alteration in the expected value of sin and cos of solar time. LGTM. Can be merged!

**Review by MadhaviJain90 — APPROVED** [review]
can be merged

**Comment by SallyElHajjar** [comment]
@MadhaviJain90   
 The difference is that time was handled incorrectly in your previous PR, (error when testing). This is the mainly edited part:
         # Convert time variable to datetime objects
        times = xarray.DataArray(
            [
                # Check if the value is a Unix timestamp (numeric type)
                (
                    numpy.datetime64(
                        datetime.datetime.utcfromtimestamp(t).strftime(
                            "%Y-%m-%dT%H:%M:%S"
                        )
                    )
                    if isinstance(t, (int, float))
                    else numpy.datetime64("".join(t.astype(str)).replace("_", "T"))
                )
                for t in times.values
            ]
        )

        print("Debug - Converted Times:", times)

        # Extract hours and minutes from the datetime objects separately
        utc_hours = times.dt.hour  # Extract the hour component
        utc_minutes = times.dt.minute  # Extract the minute component

        # Convert to fractional hours (hours + minutes/60)
        utc_hours_minutes = utc_hours + utc_minutes / 60.0

        print("Debug - UTC Hours and Minutes (in fraction):", utc_hours_minutes)

**Pull request merged**